### ISSUE TITLE
Updating anaconda.org metadata

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-forge.github.io#126]
Sometimes the metadata of a package changes. For instance, it's license and/or summary. When this happens, we can update the feedstock by re-rendering. However, I'm a bit unclear on if there is a way to update the metadata on the anaconda.org.

If there is a way to do this, I'd be curious on how one goes about doing it. If this can't be done by others (e.g. requires a token), it would be nice if we could figure out a way to do this automatically (possibly as part of the team update script or similar). Though it might be nice to automate this even if there are no such difficult requirements. If there is no way to do this, it would be good if we could figure out one.

**Comment by msarahan** [comment]
I have pointed out this issue to the anaconda.org team - hopefully they can give you some insight here.

**Comment by mcg1969** [comment]
anaconda.org updates it `repodata.json` indices programmatically, upon upload of a new package. I do not believe it is possible to hotfix metadata there without force-uploading the corrected package. (The rules are different for defaults because it is hosted elsewhere and mirrored on anaconda.org)

Kale and I were talking about the logistics of doing this. Currently we do push metadata hotfixes up to defaults, so that the metadata contained in the tarballs is out of date. In most cases conda is correctly giving priority the repodata.json data; `conda index` is the one current exception I can think of, and that needs to be fixed.

The right thing to do is to update the tarballs as well, but this changes the MD5 signature, posing a problem for people who are choosing to be so careful about reproducibility that they track and verify those signatures. @Kfranz and I discussed what it would look like to handle this use case. In short, when we force-upload a new tarball, the old one would be moved aside by _appending the MD5 signature to the filename_. Furthermore, we would incorporate a _history_ into the repodata index, so that it would be known that these files exist and what their changes were.

**Comment by ijstokes** [comment]
I may be misunderstanding the proposal, however in general a package found at a particular canonical URL should never change once put into Anaconda Cloud (or Anaconda Enterprise).  If there is a problem with it then a new one (new version) can be uploaded, and optionally the old one could be removed.  But it is a bad idea to have Anaconda Cloud return two different versions of the package at two different points in time -- which is what would be possible if one could edit package meta-data after it had been uploaded, if I am understanding correctly the proposal here.

**Comment by msarahan** [comment]
I think this is less about a single package file (which I agree should not change) and more about things like the license, description, doc url, dev url, etc - primarily the "about" section.

**Comment by mcg1969** [comment]
That's right. I agree that changing the installed files merits a version bump, but metadata-only changes are, in my view, justifiable. In fact there are even situations where metadata should be, even needs to be, hotfixed.

**Comment by mcg1969** [comment]
And @ijstokes, for the record, we do metadata hotfixes and force uploads, for better or worse, in the defaults channel. Thus it simply isn't the case in practice that the URLs (as we currently offer them) are "canonical". The proposal Kale and I discussed above will help us rectify that, but we do so _not_ by requiring that the `package-version-build.tar.bz2` URL itself be canonical. Rather, every uploaded package would be stored in a filename with its MD5 hash appended; e.g., `package-version-build.tar.bz2-hash`. _That_ would be a canonical URL. Then `package-versio-build.tar.bz2` would alias to the latest revision of that package.

**Comment by jakirkham** [comment]
My concern was license and URL data being fixed as @msarahan correctly deduced.

Though there are definitely use cases for hotfixing dependency *pinnings* of packages. See this issue ( https://github.com/conda-forge/conda-forge.github.io/issues/111 ) where @183amir proposes this exact solution recently in a discussion with @pelson.

**Comment by mcg1969** [comment]
Yep. There are several scenarios one can consider where hotfixing the dependency list is important.

**Comment by 183amir** [comment]
I am having a lot problems with a lot of packages that are not pinned! I have tried deleting some of them from anaconda.org but I think the best solution is change their metadata.
How do we feel about this?

**Comment by 183amir** [comment]
Like this one:
```
conda execute conda-forge.github.io/scripts/list_deps.py conda-forge --package bob.blitz --dependencies boost 
Fetching package metadata: ..
bob.blitz 2.0.8 np110py27_0 None   : boost
bob.blitz 2.0.8 np110py27_1 None   : boost
bob.blitz 2.0.8 np110py27_2 None   : boost
bob.blitz 2.0.8 np110py27_3 None   : boost
bob.blitz 2.0.8 np110py27_4 None   : boost
bob.blitz 2.0.8 np110py27_5 None   : boost
bob.blitz 2.0.8 np110py27_6 None   : boost
bob.blitz 2.0.8 np110py27_7 None   : boost 1.60.*
bob.blitz 2.0.8 np110py27_8 None   : boost 1.61.*
bob.blitz 2.0.8 np110py34_0 None   : boost
bob.blitz 2.0.8 np110py34_1 None   : boost
bob.blitz 2.0.8 np110py34_2 None   : boost
bob.blitz 2.0.8 np110py34_3 None   : boost
bob.blitz 2.0.8 np110py34_4 None   : boost
bob.blitz 2.0.8 np110py34_5 None   : boost
bob.blitz 2.0.8 np110py34_6 None   : boost
bob.blitz 2.0.8 np110py34_7 None   : boost 1.60.*
bob.blitz 2.0.8 np110py34_8 None   : boost 1.61.*
bob.blitz 2.0.8 np110py35_0 None   : boost
bob.blitz 2.0.8 np110py35_1 None   : boost
bob.blitz 2.0.8 np110py35_2 None   : boost
bob.blitz 2.0.8 np110py35_3 None   : boost
bob.blitz 2.0.8 np110py35_4 None   : boost
bob.blitz 2.0.8 np110py35_5 None   : boost
bob.blitz 2.0.8 np110py35_6 None   : boost
bob.blitz 2.0.8 np110py35_7 None   : boost 1.60.*
bob.blitz 2.0.8 np110py35_8 None   : boost 1.61.*
bob.blitz 2.0.8 np111py27_0 None   : boost
bob.blitz 2.0.8 np111py27_1 None   : boost
bob.blitz 2.0.8 np111py27_2 None   : boost
bob.blitz 2.0.8 np111py27_3 None   : boost
bob.blitz 2.0.8 np111py27_4 None   : boost
bob.blitz 2.0.8 np111py27_5 None   : boost
bob.blitz 2.0.8 np111py27_6 None   : boost
bob.blitz 2.0.8 np111py27_7 None   : boost 1.60.*
bob.blitz 2.0.8 np111py27_8 None   : boost 1.61.*
bob.blitz 2.0.8 np111py34_0 None   : boost
bob.blitz 2.0.8 np111py34_1 None   : boost
bob.blitz 2.0.8 np111py34_2 None   : boost
bob.blitz 2.0.8 np111py34_3 None   : boost
bob.blitz 2.0.8 np111py34_4 None   : boost
bob.blitz 2.0.8 np111py34_5 None   : boost
bob.blitz 2.0.8 np111py34_6 None   : boost
bob.blitz 2.0.8 np111py34_7 None   : boost 1.60.*
bob.blitz 2.0.8 np111py34_8 None   : boost 1.61.*
bob.blitz 2.0.8 np111py35_0 None   : boost
bob.blitz 2.0.8 np111py35_1 None   : boost
bob.blitz 2.0.8 np111py35_2 None   : boost
bob.blitz 2.0.8 np111py35_3 None   : boost
bob.blitz 2.0.8 np111py35_4 None   : boost
bob.blitz 2.0.8 np111py35_5 None   : boost
bob.blitz 2.0.8 np111py35_6 None   : boost
bob.blitz 2.0.8 np111py35_7 None   : boost 1.60.*
bob.blitz 2.0.8 np111py35_8 None   : boost 1.61.*
```
Basically all the `boost` packages should be pinned on `boost 1.60.*`.

**Comment by mcg1969** [comment]
> Basically all the boost packages should be pinned on boost 1.60.*.

@183amir, why is that? If `boost 1.60` works, shouldn't `boost 1.61` work as well?

The problem here is not the unpinned versions but the *overpinned* ones. What I'm seeing here is *exactly* the kind of problem that needs to be avoided as conda-forge implements its pinning strategies. Pinning this precisely is bound to cause problems for people. 

For instance, if `boost 1.60` is already installed, `conda` is going to resist upgrading `bob.blitz` to its latest build, because it doesn't want to have to upgrade `boost 1.61` if it doesn't have to. Oh sure, conda _will_ upgrade it if you explicitly ask it to do so, and schedule an upgrade to `boost 1.61` in the process. But it will never be part of a second-level upgrade.

Dependencies need to be *as weak as possible, but no weaker*, otherwise you are going to invite conflicts like this. And instead of doing wildcard pinning, you should be considering inequality pinning; e.g., `boost >=1.60`.

**Comment by mcg1969** [comment]
Even worse: if someone already has `boost 1.59` installed, and then they do `conda install bob.blitz`, they are quite likely going to be given builds 6 or earlier, because those are the only versions compatible with their currently installed version of `boost`.

**Comment by 183amir** [comment]
Dear Michael, bob packages (at least) look for exactly that version of
boost. So even though you can install this package alongside boost 1.61, it
will crash. I have seen this crash more than one time in our builds.

**Comment by mcg1969** [comment]
Wow, that's a serious inconvenience. Well, then I certainly agree that the metadata needs to be corrected to reflect that, across all builds.

**Comment by 183amir** [comment]
Yes. But it is also mentioned in our wiki
https://github.com/conda-forge/staged-recipes/wiki/Pinned-dependencies that
boost should be pinned like that. 👍

**Comment by mcg1969** [comment]
Very good! Then it really is just a matter of what to do about retroactive fixes.

**Comment by 183amir** [comment]
Yes! I was thinking of downloading packages, changing "info/index.json" and
replacing them. Will changing that file only be enough?

**Comment by mcg1969** [comment]
I believe so, yes.

We've been talking about this question a lot internally, because anaconda.org users don't have the same flexibility we at Continuum have with the "defaults" channel. For instance, on defaults, we can hotfix _just the `repodata.json` file_ without having to change the package. That's because defaults is a static site. In contrast, anaconda.org updates `repodata.json` automatically when a package is added or removed from a channel. So the only way to do metadata hotfixes on anaconda.org is by updating the package.

Well, you might argue that we _should_ update the package, even on defaults, when we change `repodata.json`. I am inclined to agree except for one wrinkle: changing the package alters its MD5 signature. So by replacing the package you are effectively _removing_ a package from the repo that someone else might have already downloaded and is depending on. If they're attempting to achieve reproducibility down to the MD5 level, they will be unable to do so.

One thing we're considering is that instead of _overwriting_ a package when a replacement is uploaded, we _move_ the old package aside by appending its MD5 prefix to the URL. That way it is no longer available in the index, but still available for explicit request.

**Comment by jakirkham** [comment]
Sorry @mcg1969, but Boost breaks ABI just about every minor version. See this [report]( http://upstream.rosalinux.ru/versions/boost.html ). Now this doesn't always happen, but without generating our own ABI reports we cannot determine a large range that is acceptable for Boost.

In short, I disagree that this is case of over pinning.

**Comment by mcg1969** [comment]
Fair enough @jakirkham! I mean, @183amir convinced me it was necessary for `bob.blitz`, but I was unaware this was a problem primarily caused by `boost`. Thanks for the info!

**Comment by janschulz** [comment]
See also https://github.com/conda-forge/conda-forge.github.io/issues/157 -> package native libs in packages which include the version/SONAME, so that multiple versions can be installed.

**Comment by jakirkham** [comment]
Sure, we have been trying to do our homework with these. 😄 Though there may be other ones that we may need to revisit. What we have started doing (very recently I might add), is writing the pinnings that seem acceptable for the stack in one central [location]( https://github.com/conda-forge/conda-forge.github.io/blob/631e28f3d5fc2a979c937b3dd667a6349e432126/scripts/pin_the_slow_way.py#L31-L49 ). From this PRs are generated for feedstocks to make them compliant. This way we can attempt to ensure the same pinnings are applied across the stack.

**Comment by 183amir** [comment]
@conda-forge/core I wrote an initial script to hotfix dependencies, please take a look at it so that we can go forward with this as soon as possible: https://github.com/conda-forge/conda-forge.github.io/pull/170
